L1Trigger/L1THGCal/src/backend/HGCalStage1TruncationImpl.cc

```diff
@@ -0,0 +1,103 @@
+#include "L1Trigger/L1THGCal/interface/backend/HGCalStage1TruncationImpl.h"
+#include "DataFormats/ForwardDetId/interface/HGCalTriggerBackendDetId.h"
+#include <cmath>
+
+HGCalStage1TruncationImpl::HGCalStage1TruncationImpl(const edm::ParameterSet& conf)
+    : do_truncate_(conf.getParameter<bool>("doTruncation")),
+      roz_min_(conf.getParameter<double>("rozMin")),
+      roz_max_(conf.getParameter<double>("rozMax")),
+      roz_bins_(conf.getParameter<unsigned>("rozBins")),
+      max_tcs_per_bin_(conf.getParameter<std::vector<unsigned>>("maxTcsPerBin")),
+      phi_edges_(conf.getParameter<std::vector<double>>("phiSectorEdges")) {
+  if (max_tcs_per_bin_.size() != roz_bins_)
+    throw cms::Exception("HGCalStage1TruncationImpl::BadConfig") << "Inconsistent sizes of maxTcsPerBin and rozBins";
+  if (phi_edges_.size() != roz_bins_)
+    throw cms::Exception("HGCalStage1TruncationImpl::BadConfig") << "Inconsistent sizes of phiSectorEdges and rozBins";
+
+  constexpr double margin = 1.001;
+  roz_bin_size_ = (roz_bins_ > 0 ? (roz_max_ - roz_min_) * margin / double(roz_bins_) : 0.);
+}
+
+void HGCalStage1TruncationImpl::run(uint32_t fpga_id,
+                                    const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& tcs_in,
+                                    std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& tcs_out) {
+  unsigned sector120 = HGCalTriggerBackendDetId(fpga_id).sector();
+  std::unordered_map<unsigned, std::vector<edm::Ptr<l1t::HGCalTriggerCell>>> tcs_per_bin;
+
+  // group TCs per (r/z, phi) bins
+  for (const auto& tc : tcs_in) {
+    const GlobalPoint& position = tc->position();
+    double x = position.x();
+    double y = position.y();
+    double z = position.z();
+    double roverz = std::sqrt(x * x + y * y) / std::abs(z);
+    roverz = (roverz < roz_min_ ? roz_min_ : roverz);
+    roverz = (roverz > roz_max_ ? roz_max_ : roverz);
+    unsigned roverzbin = (roz_bin_size_ > 0. ? unsigned((roverz - roz_min_) / roz_bin_size_) : 0);
```

Comment:
```suggestion
    unsigned roverzbin = 0.;
    if (roz_bin_size_ > 0.) {
       double roverz = std::sqrt(x * x + y * y) / std::abs(z);
       roverz = std::clamp(roverz, roz_min_, roz_max_);
       roverzbin = unsigned(std::abs(roverz - roz_min_) / roz_bin_size_);
```

Reply:
Implemented, but without the `std::abs` in the end since it is ensured that `roverz - roz_min_` is positive from the line above

Reply:
Yes, quite probably it is redundant, but I'm always scared when I have to cast an unsigned from a difference of terms, fearing of some possible rounding effect in the difference. What about the following?:
```C++
       double roverz = std::sqrt(x * x + y * y) / std::abs(z) - roz_min_;
       roverz = std::clamp(roverz, 0., roz_max_ - roz_min_);
       roverzbin = unsigned(roverz / roz_bin_size_);
```

Reply:
That would indeed be better. 
With  `std::abs(roverz - roz_min_)`, in the case where we have negative values, it would use some random bin, while by clamping the difference between 0 and max the under/overflows are put in the first/last bin, which is a saner behaviour.
I'll implement that